### ISSUE TITLE
Hanging notes/data corruption on GPIO input with certain MIDI sources

### ISSUE THREAD

**Issue by thorr2** [dwhinham/mt32-pi#21]
Hello, I just installed this for the first time and made my own MIDI adapter for the GPIO pins using your schematic.  I hooked up an OLED display and it works properly.  Munt is missing some note-off's and it is consistent (always happens in the same places in the same songs).  I am using Monkey Island 2 for example on my MiSTer and the music always has hanging notes in the same places that should not be there.  The display shows these notes hanging.  I have tried different chunk sizes (32 and 2048) and set my frequency to 32000 and the scaler to none.  Nothing so far seems to help.  Any ideas what could be causing this?  I have also played Doom on my MiSTer with my Roland SC-55 mk2 and it works fine with no hanging notes, so that is not the cause.  I also tried playing to the mt32-pi from my PC and it also has hanging notes.  Thanks for your help.

**Comment by dwhinham** [comment]
Hi!

Which model of Raspberry Pi are you using?
Is there any chance you can get me a recording of the hanging notes or tell me a specific song/time to look for so I can try to recreate the problem?

**Comment by thorr2** [comment]
Hello, I am using a Pi 4 model B.  Pretty much any song I play has the
problem.  I think the songs sound the same wrong way in the same places
every time I play them.  I have about 3 to 4 inch wires to the midi jack
and to the Pi GPIO pins from my midi breadboard.  I don't know if that
matters.

I tried installing Raspian with Munt to see if it was a hardware problem,
but I don't know how to get it to use the GPIO midi in port.

I will follow up with a link to a midi file and a recording of how it
sounds.

Thanks!
Michael

**Comment by thorr2** [comment]
Hello,

Here is a midi file and a recording of it.  Notice the obvious stuck note
(there are probably others).  Right at the end of the recording the note
gets unstuck.  It always sounds the same.  When I play the midi file with
Windows sound instead of through the mt32-pi, that note is not stuck.

https://www.sendspace.com/file/yzvwc8

Thank you for your help!
Michael

**Comment by dwhinham** [comment]
Thanks for sending me the MIDI.

I've just tested playing this song from Windows on `mt32-pi` v0.3.1, using my GPIO midi interface - it worked just fine with no hanging notes.

I'm guessing there is an electrical issue with the GPIO circuit.

- Can you measure the resistors to make sure they're the correct values and in-spec?
- Do you have another optocoupler you could try?
- Is your PSU and USB power cable good enough for the Pi? If you test with Raspberry Pi OS/Raspbian, you should not be getting any 'lightning bolt' symbols on the screen or undervoltage warnings in `dmesg` if your PSU is good and the Pi is under stress.

**Comment by nswaldman** [comment]
I'm experiencing something similar with the PI-MIDI board I've just received: the onboard MIDI interface is doing drunk renditions of the MIDI data while USB is fine. Here's some listening material: https://www.dropbox.com/s/5kcjc7rlauv0scf/recordings-mt32pi.zip?dl=0

It's not identical, but there's a slim chance we're both having the same problem. I'm still in the process of ruling out the usual suspects (cables, power adapters, different Pi units).

**Comment by thorr2** [comment]
Thank you for your help.  It is good to know I may be able to fix it.  I
have extra parts and can make a new midi board.  I will do more testing and
listen to and try the other person's example that is also having problems.

**Comment by thorr2** [comment]
I listened to Nuey San's file and it seems to be missing notes.  I also
think I experienced this at one point.  It's possible there is a loose
connection or bad resistor or something.  I have a couple of questions:

- Are you both trying on a Raspberry Pi 4?
- Nuey San, when you play the file, does it sound identical every time you
play it, where it always misses the same notes?

Thanks!

**Comment by thorr2** [comment]
Hello, I completely built a new MIDI interface with all new parts and I used a different set of resistors from a different manufacturer and measured each one before soldering it in to get the closest I could to the recommended values (and they were very close), and after everything was in place, I tried it and I still have the exact same problem.  I had this problem before and after installing a PCM5122 hat.  Are you testing on a Raspberry Pi 4 when you are having no issues?  Any further help would be appreciated.

**Comment by thorr2** [comment]
I noticed this diagram uses different resistor values (220, 470 and 1K).  I am not sure if trying that would help.  https://www.instructables.com/id/PiMiDi-A-Raspberry-Pi-Midi-Box-or-How-I-Learned-to/

**Comment by dwhinham** [comment]
I'd use that circuit diagram with caution, it has a pull up going to the +5V and the UART on the Pi is only capable of 3.3V, so there is risk of damage. Someone in the comments on that page also noticed this.

Sorry you're still having trouble, I'm really not sure what to suggest 😥

Can you provide hi-res photos of your circuit and how it's connected to the Pi?

The circuit in the README is the same as on this page: https://www.samplerbox.org/article/midiinwithrpi
I've tested it successfully on Pi 3, Pi 4, both constructed on a breadboard and on an assembled PCB.

Was your SD card 100% clean when you installed `mt32-pi`? (i.e. not just copied over an old Linux installation)

Can you try to test the interface in a clean install of Raspbian to try and rule out `mt32-pi`? You'll need to do the config dtoverlay changes and use ttymidi with something like MUNT or FluidSynth as suggested in this pagehttps://zuzebox.wordpress.com/tag/midi/

**Comment by thorr2** [comment]
Thanks.  It is really late, so I will provide the pictures and work on this more tomorrow.  The SD card was clean formatted before installing the MT-32 Pi.  I noticed the Diode is different in the link you provided.  I am not sure if that makes a difference or not.  Also, I have a Pi 3 that I can try it on, but it will take some work to get to it (it's in a NesPi case acting as my Pi-hole).

**Comment by dwhinham** [comment]
Uh oh - yes, you're absolutely right, it's meant to be a 1N4148.

Please try with one of those if you can!

**Comment by thorr2** [comment]
I will have to order some.  Hopefully that will fix it!  Thanks!

**Comment by dwhinham** [comment]
I hope so! 🤞🤞

If this was the reason, I apologise for the mistake - you're the first person to actually attempt to build the circuit from the schematic listed here and report back.

I'll fix the schematic ASAP.
Thanks!

**Comment by nswaldman** [comment]
@thorr2 to answer your questions:

I've tested on a Pi3 and Pi4, the dropped/stuck notes are different for each Pi tested, but they seem consistent.

**Comment by thorr2** [comment]
Sounds like we have the same problem.  I will hopefully have the new diode
tonight.  If it fixes it, maybe your pi hat also has the wrong diode.  I
will keep you posted.

**Comment by thorr2** [comment]
I installed the new diode and it did not change anything unfortunately.  I guess I will try your suggestion of installing Raspian and trying to make that work.  I tried once before but didn't get very far.  Hopefully the link you provided will provide all the necessary information.  I also have a Pi 3 I can try, so I may try that first.

**Comment by thorr2** [comment]
I tried the Pi 3 and it also had the issues, but less so.

**Comment by thorr2** [comment]
This is the spec sheet for the 6N138 I used: https://media.digikey.com/pdf/Data%20Sheets/Lite-On%20PDFs/6N138-39%20Series.pdf
Is it possible this requires 5V and I need to buy a different one that works at 3.3V?

**Comment by thorr2** [comment]
This is interesting (see the answer talking about 5V): https://electronics.stackexchange.com/questions/165255/midi-in-many-circuits

**Comment by dwhinham** [comment]
Sorry the diode didn't change anything 😢 

From the datasheet, your opto should be fine with 3.3V, which to me means that that comment ("6N138 needs a 5 V power supply") does not apply here:
![image](https://user-images.githubusercontent.com/5890747/90629316-10129580-e217-11ea-8a9f-4363a077783a.png)

It supports -0.5 to 7V, and whatever you put in will be the same as what comes out, hence 3.3V into the opto means 3.3V out, safe for the Pi.

- You should still try Raspbian.
- You could try powering the opto via 5V as per the circuit in that StackOverflow answer, but you *must* add a voltage divider so that the signal is brought down to 3.3V before going into the Pi's Rx pin or you'll permanently damage the Pi.
- Try other manufacturers of opto aswell. I'm assuming you've ruled out the opto being faulty by replacing it with another unit?

**Comment by thorr2** [comment]
Thanks.  I soldered in the 0.1 uF capacitor and changed the circuit to wire it up to the 5V on pin 8 with pin 6 going to 3.3V through the resistor per the diagram I linked.  It unfortunately did not change anything.  I used a different 6N138 when I made my second MIDI board but they were both from the same manufacturer.  I want to try Raspian, but it sounds like I will need to figure out all the steps from different sources.  I have a Pi 4, and that post was 2017.  Getting Munt to work with the GPIO MIDI IN is the challenge.  If I can't figure it out, maybe you can tell me all the steps in a list of what to do from scratch.  I can install Raspian from the imager tool and I am very familiar with Linux.
Thanks again,
Michael

**Comment by dwhinham** [comment]
No problem, I'll do a quick write-up of how to test it in Raspbian shortly.

I've noticed another mistake - in my breadboard diagram, the connections from the MIDI DIN plug are reversed, and it's because the tool I used to draw it (Fritzing) has an incorrect pin mapping for the DIN plug pins - the schematic is OK though.

I doubt this will be affect you though, as you mentioned soldering and I guess you followed the schematic and not the breadboard. I'm still interested in seeing how you've built your interface!

**Comment by thorr2** [comment]
Thank you very much.  I would really appreciate the write-up for what all to do in Raspian (including setting up Munt and using it with the GPIO midi port).  For the MIDI DIN, I went by the numbers on the schematic and the numbers embedded in the plastic on the MIDI connector.  I have uploaded pictures, so maybe you can check if it is right.  Here are the pictures:
https://www.sendspace.com/file/7kcr3i

Those are from my first midi interface that I made before changing it to 5V and changing the diode.  My second interface is soldered to the pins on the Pi Hat because I am trying to make it fit into a case and there is not enough room for the push-on connectors. They are basically the same though.

Thanks again!
Michael

**Comment by dwhinham** [comment]
No worries, I'm glad you were able to follow them!

Unfortunately I don't know what else to suggest; my electronics knowledge only goes so far and I've just been using designs that others have shared. But at least we know for certain that it's a hardware issue in your case.

Hoping someone might chime in with a suggestion or feed back their own experiences, or maybe it's worth asking in that subreddit - I'd love to hear if you get any further!

**Comment by thorr2** [comment]
Thanks!  My parts should be here by Friday.  I have two methods to try so
hopefully one or both will solve the problem.  I will keep you posted.

**Comment by icb-** [comment]
> So, I guess the next step is to order a new 6n138 chip or try an alternative chip like the H11L1.

I would strongly suggest using the H11L1 (e.g. the [ON Semi H11L1M](https://www.onsemi.com/pub/Collateral/H11L3M-D.PDF)) over the 6N138. It has an open-collector schmitt trigger output, so still needs the pull-up resistor, but unlike the 6N138, the detector side is specified for a 3v to 15v supply voltage, so you can power it from the 3.3v supply without having to use a resistor divider to get the output to acceptable voltages. If you do change to the H11L1M, keep in mind that it's not pin compatible with the 6N138. Cost is a little higher, [91¢ for the On Semi H11L1M at Digi-Key](https://www.digikey.com/product-detail/en/on-semiconductor/H11L1M/H11L1-MQT-ND/284866) at qty. 1 compared to [81¢ for the Lite-On 6N138 at Digi-Key](https://www.digikey.com/product-detail/en/lite-on-inc/6N138/160-1795-ND/1969179), but negligible if you're only buying one.

The -0.5v to 7v for the 6N138 quoted above is misleading. That's from the "absolute maximum" section of the data sheet, that lists the voltages that can be applied without _damaging_ the chip, not the voltages it works at. A traditional 6N138 is specified to work with a 4.5v to 5.5v supply. Some chips may work from a 3.3v supply, but it's working outside its intended voltage range.

**Comment by dwhinham** [comment]
Thanks for this info and clarification on my misleading interpretation of the datasheet. I couldn't find any better information to help @thorr2 on my own - sometimes, the only way to get the right answer on the Internet is to say the wrong answer so that someone will come along and correct you. 😄 

I guess 6N138s that work at 3.3v are just from blind luck. It'd be great to have a better example schematic up here based on the H11L1, so I'll order one too shortly and maybe we can improve the design.

**Comment by thorr2** [comment]
Thank you Ian.  I have an H11L1M on the way so I will be able to try it.  I
did try running the 6N138 at 5v and pulling the output down to 3.3v and it
didn't help.  Hopefully the H11L1M will solve the problem.  Thanks again!

**Comment by icb-** [comment]
Reading data sheets is a bit of a learned skill. They're often not as straightforward as they should be.

The [MIDI Electrical Specification](https://www.midi.org/specifications-old/item/midi-din-electrical-specification) allows for either 5v or 3.3v (added in 2014) from the transmitter. The recommended receiver circuit (ibid) includes a plain signal diode for reverse voltage protection, a current limiting resistor (as you would use with any LED), and a PC900V or 6N138 optoisolator (1). (The PC900V is a clone of the H11L1.) Since both optoisolators have open-collector (3) outputs, they also need a pull-up resistor.

Now that you know how the input is driven at the MIDI jack, you can look at the input specification for the optoisolator. The 6N138 [specifies](https://media.digikey.com/pdf/Data%20Sheets/Lite-On%20PDFs/6N138-39%20Series.pdf) 1.1v to 1.7v at 1.6mA (it also says 0.5mA on the first page, but doesn't specify the voltage requirements at that current) in the "Electrical Specifications" section. The H11L1 [specifies](https://www.onsemi.com/pub/Collateral/H11L3M-D.PDF) 0.75v to 1v at 0.3mA. So either one is perfectly acceptable on that front.

The MIDI specification says "The receiver must require less than 5mA to turn on." so both parts are acceptable there too. You need to select a current limiting resistor that meets that 5mA limit, and the 1.6mA for the 6N138, or the 0.3mA for the H11L1 minimums. There are also two current limiting resistor on the transmit side, each on the order of 220Ω for a 5v output, or 10Ω for a 3.3v output. Take those two resistors into account when calculating your current limiting resistor. In practice, you can probably pull more than the 5mA from a MIDI output, but don't push it too hard.

The MIDI specification also says "Rise and fall times should be less than 2 microseconds.". The 6N138 doesn't seem to specify its rise or fall time, but the MIDI specification says it has "been found to be acceptable." The H11L1 specifies 0.1µS.

Once you're past the "input" side of the optoisolator, you need to look at _your_ input's requirements. The 6N138 doesn't explicitly state its "output" voltage requirements in most data sheets, but you can get an idea of what it wants from the "test condition" column in the "electrical specification" section, that states the tests are at 4.5v. That makes it unacceptable for the 3.3v input on a Raspberry Pi. It **might** work from a 3.3v supply, but it might not. The H11L1 lists the "operating voltage range" of 3v to 15v in the "electrical characteristics" section.

MIDI runs at 31.25kbps. The 6N138 doesn't specify its switching time, but the MIDI specification says it's acceptable. The H11L1 specifies 1MHz, which is well above the needed baud rate.

Since the output of both optoisolators are open-collector, you need a pull-up resistor. MIDI is slow enough that you don't have to worry too much about the value of the resistor. If you use too high of a value, your rising edges will be too slow.

The shield and pin 2 (ground) on the MIDI connector should only be connected on the sender. Leave them unconnected on the receiver, or connected through small value capacitors to ground.

The [schematic posted above by @thorr2 from the TI forums](https://e2e.ti.com/support/microcontrollers/msp430/f/166/t/584083) is a good one to follow.

1. An optoisolator is essentially a LED and a photodiode or phototransistor in a single package. The 6N138 uses a darlington pair (two cascaded NPN BJTs) to amplify the signal. The H11L1 uses a Schmitt trigger (2) on its output stage. They're both open-collector. Being built of an LED and BJTs, the 6N138 is a current-mode analog device. That's not strictly what you want, but when driven with a logic signal, it gives a good enough logic output. The H11L1 has a proper logic output because of its Schmitt trigger.

2. A Schmitt trigger is a comparator with hysteresis. That is, it compares two voltages (in this case, the voltage present on a photodiode and an internal reference threshold) with hysteresis. Hysteresis means it has two threshold voltages. Between those two thresholds, the output will stay the same until the input crosses the other threshold. This stops spurious (unintentional) changes in output when the input is hovering around one of the thresholds.

3. An open-collector output can only pull to ground. You need a pull-up resistor to pull the output high. This is often used by multi-drop busses like I2C to prevent damage when two parts try to drive different logic states.

**Comment by thorr2** [comment]
Thank you Ian!  In the schematic I posted, are all of the resistor values
shown ok for 3.3V?

**Comment by icb-** [comment]
For the circuit on the TI forums post? Yes, the MSP430 it's drawn for is a 3.3v device too.

The resistor (R2) in parallel with the input reverse polarity diode (D1) can safely be removed. If you're trying to cut down parts, the reverse polarity protection diode can be removed with the worst that might happen is blowing out the diode in the optoisolator if you miswire the MIDI input (but not damaging anything past the optoisolator. That's what it's designed to protect.) The shield capacitor (C1) is completely optional, just reducing EMI. It will probably also work without the bypass capacitor (C2) on the optoisolator.

A minimal circuit is the XLR connector, the current limiting resistor (R1), the optoisolator, and the output pull-up resistor (R3).

Also note that the circuit shows two different grounds. Chassis ground (the 3 prong "fork" facing down) on the shield capacitor (C1), and signal ground (the downward pointing triangle) on the MCU and optoisolator. They would generally be tied together in your circuit

**Comment by thorr2** [comment]
Great!  Thank you very much.  I will keep everyone posted how it goes.

**Comment by thorr2** [comment]
I installed the new MIDI circuit (minus the capacitor on pin 2 of the midi connector going to ground).  When I installed the mt32-pi onto the micro SD card, the display shows pt32-pi v0.31 --> Init mt32emu (I2S) and stays there.  I am not sure what could be wrong at this point.  I will try installing Raspian again and see if I can get it to work.

**Comment by thorr2** [comment]
I got it to work in Raspian.  I assume the mt32-pi code is trying to auto-detect the GPIO midi interface and doesn't like the new circuit for some reason, but it works in Raspian.  However, with the completely new circuit using the H11L1M chip, I STILL have the same problem with the same hanging notes.  The only thing I can think of is it doesn't like the 4 inch wires between the midi jack and my midi circuit and/or the 4 inch wires between my midi circuit and the GPIO pins.  There is nothing else to replace.  Or, there is some kind of software issue, but I tried my PI 3 previously and still had problems.  I guess I have to give up and buy a USB midi interface unless anyone has any other suggestions at this point.

**Comment by thorr2** [comment]
I tried disconnecting R2 and changing R3 to 270 ohms.  It made no difference.

**Comment by thorr2** [comment]
I have a Turtle Beach USB MIDI adapter and it worked fine in Raspian when I connected it to my Roland UM-One MK2 on my PC and MIDI played perfectly.  I then put the MT32-PI software back on there and I still can't get past the Init screen.  I am not sure why it won't get past the initialization anymore.  I tried connecting the Roland UM One to the Pi and it still won't get past the Init screen.

**Comment by thorr2** [comment]
I disconnected the DAC and set it to PWM and it still hangs during the Init.  I can't even get it to work with USB at this point unless I use Raspian.  This is so frustrating. :-(

**Comment by thorr2** [comment]
I forgot to put the MT32 rom files on there which is why it was getting stuck at the init screen.  Now I have the problem that the notes are cutting off too early when using USB.  I don't know if it is due to the hacked together temporary MIDI coupler I made or the Turtle Beach adapter.  I don't remember having this problem on Raspian.  I will order a proper coupler and go from there.

**Comment by thorr2** [comment]
Swapping the Turtle Beach and the Roland UM one MK2 fixed it.  The Turtle Beach is plugged into the PC and the Roland is plugged into the Pi and it works without dropping the notes.  So I finally have a working solution.  I have no idea why I had problems using the GPIO MIDI input but at least this works.  Thanks for your help everyone.

**Comment by dwhinham** [comment]
@icb- 
Thankyou for taking the time to write the extremely detailed replies with the MIDI electrical explanation - this is invaluable information. I still want to correct the sample schematic in the README, so I have received an H11L1 and will be testing it at some point over the weekend.

@thorr2 
> I forgot to put the MT32 rom files on there which is why it was getting stuck at the init screen.

You've just reminded me that I need to write an error message mechanism for the LCD!
As for MIDI, there's no detection for GPIO MIDI. The logic is: if `usb = on` in the config, scan for USB MIDI devices, pick the first one and try to enable it. If none found, assume GPIO. `mt32-pi` will not hang because of a missing MIDI interface, it'll just assume you have a GPIO setup and will wait for something to start talking on the RX pin.

So - final outcome is no GPIO circuit worked? What about when the Roland interface is sending to the GPIO circuit - did you try that? USB MIDI interfaces can definitely be flaky as we've seen in other issues.

**Comment by thorr2** [comment]
Thanks.  The final outcome is every GPIO circuit sounded wrong in exactly
the same places leading me to believe something else is causing the issue.
It might be a timing issue or something like that.  I was using the Roland
interface the whole time until I started trying to use USB because I needed
two interfaces for that.  The Roland interface is the better of the two.  I
also tried it from two sources, my PC and my MiSTer, and both were fine via
USB and both were wrong but sounded the same exact wrong way every time I
played the same song via GPIO MIDI.

An error message on the screen would be great for the missing roms.  Will
it work with cm32l roms?  With what filenames? That was not talked about in
the documentation.

Thanks again!

**Comment by dwhinham** [comment]
If you want to log what `mt32-pi` is receiving, follow this:
https://github.com/dwhinham/mt32-pi/issues/20#issuecomment-671571547

Run this build and play Pink Panther, then move the log file off the SD card, and then repeat a few more times.
Attach the log files here.

If we use a diff tool (eg. VS Code), it will prove whether the stuck notes are happening in the same place or not.

I'm sorry to say, but I doubt there is anything I can do here at all - I can't reproduce it, others are using GPIO just fine too, and the behaviour is similarly broken on an entirely different environment - everything points to some hardware issue/incompatibility with your particular setup or combination. There's hardly anything that can go wrong on the software side - `mt32-pi` simply feeds what's coming in on the serial port to the synthesizer.

`mt32-pi` should work fine with CM32L ROMs - anything Munt supports, we should support. Filenames stay the same though, the contents of the ROM and its hash is what changes the behaviour.

**Comment by glaucon1984** [comment]
Hi,

I received my PI-MIDI from Aranet (not sure if I should put a link) and I set it up, I can confirm the issues from the first post. I can also hear issues with Monkey Island 1, the intro starts with a couple of sustained notes that end too early.

I've connected the MIDI output to a real Roland CM-500 (in CM-64 mode) and it works flawlessly, but MT32-PI seems to have issues.

I thought it might be my PI-MIDI board, but I'm wondering if there is any issue at software level. Can anybody else test the intro of Monkey Island 1 on MT32-PI and compare with one of the many videos in Youtube?

**Comment by dwhinham** [comment]
This video on Arananet's channel is actually recorded by me: https://www.youtube.com/watch?v=w93uoiFhk-Y
I will record another video with the full Monkey Island 1 intro on my own Pi-MIDI shortly.

If you get the same faulty behaviour after following this troubleshooting guide (https://github.com/dwhinham/mt32-pi/wiki/Troubleshooting) then you should contact Arananet for a replacement.

**Comment by dwhinham** [comment]
I constructed the circuit from https://e2e.ti.com/support/microcontrollers/msp430/f/166/t/584083 on breadboard, but **without** C2 and R2 as I didn't have them to hand. Please excuse the IO expander on the right, I'm trying to prototype control panel stuff and I only have one breadboard!

![DSC00019](https://user-images.githubusercontent.com/5890747/91595728-ec102c00-e95b-11ea-9491-49271c5177dd.JPG)

It works perfectly, just as well as the 6N138s I've been using up until now - no hanging notes. I could made another video but there's really nothing to see. Again, source is the USB Uno and ScummVM from the PC.

(@icb- I understand that C2 is probably for decoupling/noise suppression, but what purpose does R2 serve here? I haven't seen it in any of the other MIDI circuits - is it important?)

**Comment by Braincell1973** [comment]
I have a data point midi board wired to the gpio, I'm getting major amounts of missed keypresses from 2 different midi keyboards, using 1 of the keyboards in USB midi mode worked perfectly its just the gpio thats wonky :(

**Comment by glaucon1984** [comment]
Hi, I've done more research, and the PI-MIDI hat works fine using a real DOS machine. So the situation is quite surprising: MT32-PI+PI-MIDI works fine on a real DOS Machine, and MiSTer FPGA works fine on a real Roland device, but MiSTer FPGA with PI-MIDI+MT32-PI doesn't play well together.

MiSTer FPGA reaches a limit that real Roland hardware can tolerate but not this solution, and a real DOS machine with a proper MPU-401 speaks correctly enough to this solution that there are no issues, somehow **MT32-PI and MiSTer FPGA are not compatible between them.**

I guess this is such an edge case that it will be impossible to troubleshoot. I haven't tried your troubleshooting guide but I wonder if it's worth the effort.

**Comment by dwhinham** [comment]
> I have a data point midi board wired to the gpio, I'm getting major amounts of missed keypresses from 2 different midi keyboards, using 1 of the keyboards in USB midi mode worked perfectly its just the gpio thats wonky :(

Hmm. Is this an `mt32-pi` issue though? Does it work properly under Linux?

> Hi, I've done more research, and the PI-MIDI hat works fine using a real DOS machine. So the situation is quite surprising: MT32-PI+PI-MIDI works fine on a real DOS Machine, and MiSTer FPGA works fine on a real Roland device, but MiSTer FPGA with PI-MIDI+MT32-PI doesn't play well together.

That's... odd, and interesting!

> MiSTer FPGA reaches a limit that real Roland hardware can tolerate but not this solution

What kind of limit? Sorry - can you explain?

> and a real DOS machine with a proper MPU-401 speaks correctly enough to this solution that there are no issues, somehow **MT32-PI and MiSTer FPGA are not compatible between them.**
> 
> I guess this is such an edge case that it will be impossible to troubleshoot. I haven't tried your troubleshooting guide but I wonder if it's worth the effort.

I was sent this video of `mt32-pi` being used with a MiSTer - is their setup different to yours?
https://www.youtube.com/watch?v=o7yOY2bKXfk

Unfortunately I don't own a MiSTer and I don't fully understand all of the hoops they're jumping through to get MIDI out of the device - it's going to be very hard for me to help in this case.

**Comment by Braincell1973** [comment]
I have a mister but I don't know the connection method so I can test that .

I think I'm going to slap a scope on the output from the data point although its puzzling that the usb works ok and several people have troubles with the gpio.......

The midi circuit is passive  being just a signal conversion so I think its unlikely its the datapoint board........

**Comment by dwhinham** [comment]
> I think I'm going to slap a scope on the output from the data point although its puzzling that the usb works ok and several people have troubles with the gpio.......

An oscilloscope capture would be extremely useful, please feel free to share your findings. I'd be interested in seeing the waveform at the Rx pin.

I wish I had an oscilloscope so that I could see what my (working) setups look like.

**Comment by Higgy69** [comment]
Hi,
I also just got my PI-MIDI. 
I was playing Monkey Island 2 (SCUMMVM on WinPC) and sounded good. I then tried AMIGA, sounded good (although not heard MIDI on AMIGA so no reference) but when you exit the game the last note played continues to play. Resetting AMIGA does nothing, you have to Power Cycle the Pi.
I was logging in to mention this when I read about Money Island 1, so I tried that and yes I am getting hanging notes!?

    Raspberry Pi 3 1.2GHZ
    UM-1EX Roland USB from PC running ScummVM
    mt32-pi v0.3.1 (latest release version)

Could it be the USB -> MIDI interface?
Also there are quite a few config options. I have tried 'Good' & 'Best' but no difference on Monkey Island 1 with hanging notes.
I also have a MiSTER but would really need a HD image for the A0486 core ready to boot with MIDI to test. @Braincell1973 - Once in the Core goto Core Menu and press right, there is URLINK or somthing like that, scroll through connection until MIDI is selected. My USB MIDI interface flashed and I could see the MiSTER connected on Pi-MIDI OLED.

Hopefully we will get there :) @dwhinham - your Money Island 1 sounds fine with PI-MIDI, so what are the variables:
USB MIDI
config file
Pi version (video is using a 4, do you have a RPi 3 1.2GHz? )
PSU? I am not sure. If it works on other RPi setups why would PI-MIDI be any different. Can't be a lot of extra current draw?

**Comment by glaucon1984** [comment]
> What kind of limit? Sorry - can you explain?

I have no idea, maybe talks too fast for the Raspberry Pi, maybe it's too slow, I wish I had more knowledge about the topic. I have tried with and without SoftMPU, I have tried with and without "delayedsysex", it doesn't work well.

> I was sent this video of `mt32-pi` being used with a MiSTer - is their setup different to yours?
> https://www.youtube.com/watch?v=o7yOY2bKXfk

I'm not sure what USB-MIDI adapter are they using on MiSTer side, I'm using a Roland UM-One.
He's testing General Midi in Duke Nukem 3D, I've only tested native MT-32. I'd say the strange audio on Duke Nukem 3D soundtrack is caused by the mapping of MT-32 to General Midi.
I'm also using 31250 bauds.

The funny thing is that the MiSTer setup has an ARM CPU besides the FPGA and it has a mode that supports running Munt locally, but the notes sometimes don't come at the correct tempo, which we blame on the lack of CPU power, so I wanted to try an external solution that is small and easy to carry as the MiSTer.

**Comment by Braincell1973** [comment]
I'm using a pi3a and datapoint in a custom 3d printed case with oled. The way I see it there are only a few variables affecting gpio...

Drive current affecting opto output.........
Opto speed......
Output signal condition.......
Mt32pi software ......

As there are clearly working gpio solutions its unlikely to be mt32pi, really leaning towards the midi interface output to gpio...........

**Comment by dwhinham** [comment]
@Higgy69 I hope we'll get there too, I feel a bit useless here because I'm seeing all these failure reports but I don't know how I can help without being able to reproduce it... 😞 

> config file

Only things I changed were the minimum to get Pi-MIDI fully working: `output_device = i2s` under `[audio]` and `type = ssd1306_i2c` under `[lcd]`.

> Pi version (video is using a 4, do you have a RPi 3 1.2GHz? )

I only have a Pi 4 and a Pi 3 B+, the 1.4GHz version - they both work fine. However, I know that @nswaldman (who has problems) is using a standard Pi 3.

> PSU? I am not sure. If it works on other RPi setups why would PI-MIDI be any different. Can't be a lot of extra current draw?

I use a Sony Xperia phone charger (UCH32C), which does USB C power delivery; 5V 3A.
I've also successfully used a weaker PSU - probably too weak - with the Pi 3 B+ (an old HTC 1A phone charger).

@glaucon1984 
> The funny thing is that the MiSTer setup has an ARM CPU besides the FPGA and it has a mode that supports running Munt locally, but the notes sometimes don't come at the correct tempo, which we blame on the lack of CPU power, so I wanted to try an external solution that is small and easy to carry as the MiSTer.

Ah yes, I was told about this. Munt is *very* CPU hungry, you definitely need over a GHz of ARM to reliably render all of the MT-32's polyphony properly, and I guess the MiSTer's auxiliary CPU just isn't quite enough.

**All:** I've started a wiki page here: https://github.com/dwhinham/mt32-pi/wiki/Compatibility-reports

It would be very helpful if you can add your results to it with as much detail as possible.

Thanks!

**Comment by Braincell1973** [comment]
This is interesting I'm using rom....

Mt32_control.1988-09-30.vx.xx-patched.rom

Will try with the 1.07 rom tomorrow........

**Comment by Higgy69** [comment]
@dwhinham - as you have a PI-MIDI a number of us should really be able to get to the same hardware setup as you.
Just tried with my RPi 3B+ 1.4Ghz and still have the hanging notes. Monkey Island 1 intro is really sounding bad compared to your video.  I am also now using an official RPi PSU. So I think we can rule out PSU.
Here is my editable config options in case something is different:

usb = on
output_device = i2s
sample_rate = 96000
chunk_size = 512
i2c_dac_address = 4c (00 also seems to work)
i2c_dac_init = none
resampler_quality = best
type = ssd1306_i2c
width = 128
height = 32
i2c_lcd_address = 3c

I am using the NON patched Control ROM and confirmed SHA1 with yours.
So what about USB - MIDI device is that now the main area of difference to focus on? Could there would be Windows variances? I am on Win10 64bit (have options of Win7 32bit)
My MIDI-USB is plugged into monitor's USB hub, pretty sure it is USB 2.0 as I use if for USB + MicroSD file transfer and get high MB/s.
My USB-MIDI (UM-1EX) is being used in ADVANCED DRIVER MODE - OFF This means it uses standard Windows Drivers. I can install Roland drivers but it is not needed after something like WinXP.
Does the MicroSD card do any buffering/access?  I am using an old (probably slow) 2GB formatted in FAT32.

**Comment by Braincell1973** [comment]
For me I was using 2 different midi keyboards and playing notes by hand , both keyboards with midi over gpio can't get above 1 note per second ish before next note is missed,  using 1 of the keyboards in usb mode directly connected to the pi worked flawlessly........

**Comment by dwhinham** [comment]
@Higgy69 

`i2c_dac_address` and `i2c_dac_init` have no effect on Pi-MIDI; it's using a DAC chip that needs no init sequence, so you can ignore these. completely.

What I can also tell you is that Pi-MIDI is using a circuit design that powers a 6N138 with only 3.3V (refer to the discussion with @icb-  re: electrical specs) way back up in the conversation. So that's a factor to consider (the chip is being operated out of spec).

As a side note, I have recommended that Arananet temporarily cease production of the Pi-MIDI until this issue is resolved, or at least that he build new units using H11L1s instead. Though please note that I am not in any way involved in the design, production or sales of this board - I can only try to provide software support.

Regarding Windows - standard Windows 10 here, USB Uno is using basic Microsoft built-in MIDI driver; nothing special.

SD card: The card is only accessed by `mt32-pi` at startup - to load the config and ROMs. After that there is no further disk access at all, so if it's attempting to play sound, it's successfully started up and the SD card can be taken out of the equation.

@Braincell1973

I recommend anyone with misbehaving MIDI interfaces to get a keyboard, or at least a **software keyboard** such as [Pocket MIDI](https://www.morson.jp/pocketmidi-webpage/). I've mentioned this in the [troubleshooting guide](https://github.com/dwhinham/mt32-pi/wiki/Troubleshooting):

> You could use a Windows PC with USB MIDI interface and Pocket MIDI to send MIDI notes by playing an on-screen keyboard. This way you can ensure that all key presses result in a corresponding Note On and Note Off pair. Try playing keys rapidly - if any Note Offs are missing, you probably have a hardware issue and missing/hanging notes will occur.

**Comment by Braincell1973** [comment]
OK.... so changed rom to 1.07.

Hardware ..... 
Midisport 2x2 , win 7 scummvm 2.1.2
Pi3a and datapoint 6n138 board
128 x 64 oled
Pi pwm output

Working flawlessly in sq1, sq5 , dott.........

![15986931727887310985884491634335](https://user-images.githubusercontent.com/29205818/91633663-245a4d80-e9e2-11ea-8b75-de18647597cf.jpg)
![15986932247774440615589383053790](https://user-images.githubusercontent.com/29205818/91633671-33410000-e9e2-11ea-9889-106ad2f37a76.jpg)

**Comment by icb-** [comment]
I don't have the hardware to test myself, but I think it would be worth checking the return code from the serial `Read()` call. Read errors are being checked, but if there are any errors, [the whole buffer (up to 1024 bytes) gets dropped](https://github.com/dwhinham/mt32-pi/blob/fad824a2310c410f633d2213d4022f211e55c415/src/kernel.cpp#L316-L318), and nothing gets logged. I suspect you might be seeing `-SERIAL_ERROR_OVERRUN` (`-2`) there.

The SOC in the Raspberry Pi 4 has [32 byte receive buffers](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2711/rpi_DATA_2711_1p0.pdf#page=182), so if you're not reading quickly enough, it's easy to overflow. You're using interrupts, so the interrupt needs to be serviced at least every 8mS if you have a constant stream of data coming in. The [interrupt handler](https://github.com/smuehlst/circle/blob/aa0b2a4b34214d95511a20f585506eab86d9f283/lib/serial.cpp#L465-L553) does write to a [2048 byte](https://github.com/smuehlst/circle/blob/aa0b2a4b34214d95511a20f585506eab86d9f283/include/circle/serial.h#L30) [buffer](https://github.com/smuehlst/circle/blob/aa0b2a4b34214d95511a20f585506eab86d9f283/include/circle/serial.h#L111), that gives you about 500 mS before it overruns, **if** the interrupt is being serviced quickly enough.

**Comment by dwhinham** [comment]
Good idea. 
I've made a new branch (dwhinham/mt32-pi@7a582d33c4ba2d64dad0e2a6c92ae4d240de55c5) with the following:

- Use `SERIAL_BUF_SIZE` instead of the arbitrary 1024, we may as well try and fetch the entire buffer.
- Report the error conditions into the general log and the LCD screen.

Testers with hanging notes - please try the following kernels (download the kernels artifact and extract to SD card):
https://github.com/dwhinham/mt32-pi/actions/runs/230259806

Does anything show on the LCD when the hanging notes occur?

If you don't have an LCD, or want to see more detailed logs, use these HDMI kernels instead and connect an HDMI monitor:
[kernels-hdmi.zip](https://github.com/dwhinham/mt32-pi/files/5145591/kernels-hdmi.zip)

Note that the HDMI kernel will fail to boot on Pi 4 if you **don't** connect a monitor (that's why it's not enabled in regular builds).

**Comment by glaucon1984** [comment]
I've tried the new kernels. The behaviour, for me, with Monkey Island 1 on MiSTer FPGA, is the same. The OLED screen didn't show any error.

**Comment by Higgy69** [comment]
Sorry same here. Just tried with RPi 3 with Win PC and same with Monkey Island 1. Also no error is shown on OLED.

**Comment by Higgy69** [comment]
My PC monitor also has HDMI, so I tried the HDMI Kernel and got no picture/display (RPi 3 1.2Ghz)?

**Comment by dwhinham** [comment]
Thanks for testing. I didn't thinking there'd be any overflow condition because the serial interrupts are being serviced quickly and the main loop emptying the receive buffer doesn't have much else to do apart from update an optional LCD (and people mentioned that disabling the LCD makes no difference).

Additionally, @Braincell1973 can't even get more than than a single note per second to work reliably, so we're not even coming close to the limits of the FIFO.

Still, it was worth putting the error checks in anyway to rule it out.

@Higgy69 Did you connect the display after powering on? Hotplug doesn't work with these kernels, the monitor has be connected before startup.

Cheers!

**Comment by Braincell1973** [comment]
Mine working perfectly now after moving to midisport output on Windows,  gpio is fine now

**Comment by dwhinham** [comment]
I think you're onto something here, @Braincell1973. Perhaps the key to understanding what's going on here is measuring/capturing the electrical output of a working MIDI source versus a non-working one with an oscilloscope.

Silly idea for those of you with additional MIDI gear - does putting extra devices in between the source and the Raspberry Pi with GPIO MIDI "fix" things? E.g. putting the RPi on the MIDI OUT or THRU of a keyboard, synth module, or MIDI splitter? Some synths might need you to enable something to forward MIDI data through to the OUT if it doesn't have a THRU.

**Comment by Braincell1973** [comment]
As I'm using the same 6n138 boards albeit running out of spec at 3v3 the only change is whats driving the opto...... ( and the 1.07 rom but I doubt thats it )

I will try the midisport attached to my mister next......

**Comment by thorr2** [comment]
I am currently traveling so won't be able to test until at least Wednesday
evening.  Thanks for all of your efforts.  I have a MiSTer and got it
working via USB but not GPIO, same as with my PC.

**Comment by icb-** [comment]
I think there are two places I would look next

1. Are MIDI messages being parsed correctly before handing off to `MIDIPacketHandler`? The loop in `UpdateSerialMIDI` mostly look ok, but I'm not sure if it's handing off 2-byte messages correctly. It looks to me like it always says [the message is the full length of the buffer](https://github.com/dwhinham/mt32-pi/blob/master/src/kernel.cpp#L359) (3 bytes). I think that should be `MIDIPacketHandler(0, mSerialMIDIMessage, mSerialMidiState+1)` instead (though that misuses the state machine's state parameter a bit...) That would make sense with the problem happening at the same place every time. I think `MIDIPacketHandler` is probably correct, since it's the same handler for data coming from the USB MIDI interface that works.

2. Is the baud rate being set correctly? If it's slightly off, you might get marginal results.

The code in Circle [for setting the divisors](https://github.com/rsta2/circle/blob/master/lib/serial.cpp#L255-L261) looks right to me for the [PL011 documentation](https://developer.arm.com/documentation/ddi0183/f/programmer-s-model/register-descriptions/fractional-baud-rate-register--uartfbrd). The default clock for the UART is 48MHz, so should result in IBRD of `96` and FBRD of `0`.

This could be confirmed by immediately writing read data back out through the UART and checking the output baud rate with a logic analyzer or MSO. This would also tell you if the data _is_ being read and is being lost elsewhere by comparing the data on the `IN` and `OUT` lines.

If you don't have a logic analyzer, but do have another MIDI device, you can get close by wiring up a proper MIDI output circuit to the TX pin. Newer MIDI devices should be OK with getting 3.3v on their `IN` port. You can probably pull enough current directly from the TX pin (spec says 16mA per pin, 50mA total) so you can get by without a buffer. You would connect the TX pin through a 10Ω resistor to pin 5 of a DIN5F socket, a 33Ω resistor from 3.3v to pin 4 on the DIN5, and connect pin 3 and the shield to ground. Pins 1 and 3 are left unconnected. If you want to use a buffer to protect the Raspberry Pi, or for older equipment that might not work with 3.3v, connect a non-inverting 5v buffer that can accept 3.3v input (e.g. one gate in a SN74AHCT125) before the resistor, and use 220Ω for both resistors. This is the "OUT" circuit described [in the midi specification](https://www.midi.org/specifications-old/item/midi-din-electrical-specification).

If that mirrored data is correct, then the problem is farther down in the `mt32-pi` code.

**Comment by dwhinham** [comment]
>    It looks to me like it always says [the message is the full length of the buffer](https://github.com/dwhinham/mt32-pi/blob/master/src/kernel.cpp#L359) (3 bytes). I think that should be `MIDIPacketHandler(0, mSerialMIDIMessage, mSerialMidiState+1)` instead (though that misuses the state machine's state parameter a bit...) 

Good catch! I've tried to improve this in dae2300e97e6c5063fc62383cbc244b0a223f554.

> It also looks like when you get a second status command where you expect a data byte, one of the status commands gets lost. That would make sense with the problem happening at the same place every time.

I'm not sure what the right thing to do would be in this case. This code is a leftover from the [miniorgan sample](https://github.com/rsta2/circle/blob/2832f256e3cc8c5e70d5daa5b41cac5f55eaa1fe/sample/29-miniorgan/miniorgan.cpp#L163) and I'm not sure if it's even needed. Nevertheless, I've added a quick bit of logging code here: c897d35e5b340f983f58bb28931b7a00e4679f57

On my setup, this condition never gets hit when playing the Monkey Island 1 intro at least.

> The code in Circle [for setting the divisors](https://github.com/rsta2/circle/blob/master/lib/serial.cpp#L255-L261) looks right to me for the [PL011 documentation](https://developer.arm.com/documentation/ddi0183/f/programmer-s-model/register-descriptions/fractional-baud-rate-register--uartfbrd). The default clock for the UART is 48MHz, so should result in IBRD of `96` and FBRD of `0`.

I was thinking along the same lines (baud possibly being off), so I added a quick hack to the Serial class to log the clock and we get `nClockRate` = 48000000Hz, `nIntDiv` = 96 (value being passed to IBRD), and `nFractDiv` = 0 (value being passed to FBRD), which is what you also expected, so this looks good. I'm not sure if it's worth trying to set the clock to 3MHz and using a divisor of 6 instead; some old posts around the web use this setup...

> This could be confirmed by immediately writing read data back out through the UART and checking the output baud rate with a logic analyzer or MSO. This would also tell you if the data _is_ being read and is being lost elsewhere by comparing the data on the `IN` and `OUT` lines.

Great idea again, thanks! I've added a 'mirror' feature in 72b9b18e3cad114a34e4fbae4758054c1c43f655 in anyone wants to try this.

**Testers:**
Please try https://github.com/dwhinham/mt32-pi/actions/runs/230494019
or this HDMI build: [kernels-hdmi.zip](https://github.com/dwhinham/mt32-pi/files/5146003/kernels-hdmi.zip)

We're looking for any errors being logged on the LCD or HDMI console, and any MIDI sound improvement from correcting the handling of 2-byte messages.

Also, if you'd like to try building @icb-'s MIDI OUT connector, this build should simply replay all received data directly back out of the serial port.

As an alternative to building a MIDI OUT DIN circuit on the TX pin, what I did was get a spare FTDI FT232R USB serial adaptor, and connect it to the Pi's TX pin. The FT232R can be configured to run at MIDI baud with `setserial -v /dev/ttyUSB0 spd_cust divisor 768`, which sets the clock such at using 38400 baud is really 31250. I then used `ttymidi` and Munt in the same way as in my troubleshooting wiki page (but on my PC and not a Pi), and was able to successfully hear the music from both the `mt32-pi` and from my machine running Munt receiving from to the Pi's serial output.

**Comment by dwhinham** [comment]
@Braincell1973 I've just noticed your nice MIDI I/O board has an output DIN - this is very useful and should give you what @icb- described above.

It would be very interesting if you were able to use the above build, put the "not working" USB MIDI interface into the Pi's input, and use a second interface on the Pi's output. Then you could use Munt on the PC to receive from the 2nd interface and compare the sound to what the Pi is outputting. Hope this makes sense!

**Comment by glaucon1984** [comment]
Just tried the new kernels, there are still hanging notes and similar behaviours with MI1 and MI2 intros. Still no messages on the OLED display of my PI-MIDI besides "Lucasfilms Games" and the instrument indicators.

About logging the MIDI out, I'm afraid I would be useless with a soldering iron if I had one, but MiSTer solution is based on [open source code](https://github.com/bbond007/MiSTer_MidiLink) running on Linux, so I wonder if it's possible to log the MIDI data at a software level (and if that would require involvement from the MiSTer dev)

**Comment by icb-** [comment]
> As an alternative to building a MIDI OUT DIN circuit on the TX pin, what I did was get a spare FTDI FT232R USB serial adaptor, and connect it to the Pi's TX pin and ground. The FT232R can be configured to run at MIDI baud with setserial -v /dev/ttyUSB0 spd_cust divisor 768, which sets the clock such that using 38400 baud is really 31250 (see this page for how this is calculated/how to adapt for other USB serial adaptors). I then used ttymidi -b 38400 and Munt in the same way as in my troubleshooting wiki page (but on my PC and not a Pi), and was able to successfully hear the music from both the mt32-pi and from my machine running Munt receiving from the Pi's serial output.

If you have a second Raspberry Pi, you can use the `midi-uart0` overlay to reconfigure the first UART so `38400` baud is really `31250` baud, and use `ttymidi` to inspect what's coming. If it's a Raspberry Pi 4, you can also use the `midi-uart1` overlay to reconfigure the second UART the same way, to run a second `ttymidi` to capture two MIDI streams. Check the help for `ttymidi` for the flag to attach to a different serial device.

You *can* connect the TX pin of one Raspberry Pi to the RX pin of a second Raspberry Pi, or the output of the optoisolator circuit to the RX pin of two Raspberry Pis. You just have to make sure you're not creating ground loops, by using isolated power supplies or connecting both Raspberry Pis to the same power supply.

**Comment by Braincell1973** [comment]
@dwhinham yea I can try anything but I only have the 1 midi cable....... .

From my perspective, my setup had major latency issues going from a keyboard to serial midi and as soon as I went from the midisport 2x2 everything worked fine , so as far as I can reason the hardware and mt32pi is fine , the only difference was the drive to the opto and the change of rom.......

I'm going to see if I have a 2nd cable in the loft and use the midi in out and thru on the midisport to see if it enables the 2 keyboards I tried to work.........

**Comment by dwhinham** [comment]
@icb- 

> If you have a second Raspberry Pi, you can use the `midi-uart0` overlay to reconfigure the first UART so `38400` baud is really `31250` baud, and use `ttymidi` to inspect what's coming. If it's a Raspberry Pi 4, you can also use the `midi-uart1` overlay to reconfigure the second UART the same way, to run a second `ttymidi` to capture two MIDI streams. Check the help for `ttymidi` for the flag to attach to a different serial device.

Thanks, I've got most of this documented over on the [troubleshooting page](https://github.com/dwhinham/mt32-pi/wiki/Troubleshooting#setting-up-and-testing-a-gpio-midi-interface) (thought not the `midi-uart1` info!).

@Braincell1973 Cool, thanks - any more test results will be very useful.

Cheers both.

**Comment by Higgy69** [comment]
@dwhinham - I did hot plug HDMI before, so this time I did it correct ;) and got an output.

Ok this might be something you always do, but we are not (or not all of us) - Turning on Pi power when music/midi/scummvm is playing.  I have noticed that things sound **very different** in a good way, many more notes etc, if I power on when Monkey Island 1 is running and not before. I still get hanging notes, but the sound is very different.
I was playing Monkey Island 2 and doing the powering on during music playing, it sounded good but you can see on the OLED that after a short while there is a Hanging Note on 5. It says on for a good while. 

Picture on left is MT32-pi running before starting ScummVM. Picture on right is when I power on while music is playing
**https://pichost.net/img/2JSk0**
If powering on before (left hand picture) the Top Info on OLED does not change.
Here are my HDMI screen grabs. Same again left is initial Pi power on in Windows, Right is when music is already  playing.
**https://pichost.net/img/2JVDx**

Thanks

**Comment by dwhinham** [comment]
> @dwhinham - I did hot plug HDMI before, so this time I did it correct ;) and got an output.

Great 😃 
 
> I have noticed that things sound **very different** in a good way, many more notes etc, if I power on when Monkey Island 1 is running and not before. I still get hanging notes, but the sound is very different.

That's because the synth won't have received any of the SysEx or program changes to set instrument parameters and assign instruments to parts, so you'll hear music, but all the instruments will be wrong. This setup data is often sent out just once at the start of the song.

Your logs look fine, although you should set `i2c_dac_init` to `none` to avoid the "I2C write error". Your DAC doesn't need initializing and it's not connected to the I2C bus. This is completely unrelated though and won't have any effect on the problems you're having.

I'm guessing that some of the corrupt MIDI data also got interpreted as an Active Sensing message, hence the "Active sense timeout".

**Comment by olliraa** [comment]
As advised by @dwhinham I'll report in this thread. Also experiencing these weird "notes not playing" problems. Here's a more detailed info regarding the setup:

Rpi 3b+
mt32-pi 0.3.1
PI-MIDI by Araranet

Tested with the following controllers:

Kawai K1-II (ch 2)
Roland RD-64 (ch 2)
Midi cable tested working just fine with both of the controllers and Behringer USB/midi interface + Arturia software synths.

**Comment by dwhinham** [comment]
I've just had an absolutely shocking idea while reading MIDI specs.
**Running Status**. **We don't do Running Status.**

An account of similar experiences here: http://blog.cornbeast.com/2012/12/beware-of-midi-running-status/

If the firmware of some USB interfaces "optimise" outgoing packets by using the Running Status feature, it would explain everything.

I will try to implement this as soon as possible. Please stand by. 🤞 🤞 🤞

**Comment by dwhinham** [comment]
**Testers with hanging notes**:

Please try this build:
https://github.com/dwhinham/mt32-pi/actions/runs/236578640

As usual, unzip the kernels and try as much gear as possible.

Anxiously awaiting your test results! 🤞😄

**Comment by Higgy69** [comment]
Applies virtual pat on back :)
Just tried Monkey 1 & 2 and no hanging notes ;)

Rpi3 1.2GHz
Win10 ScummVM
USB-MIDI UM-1EX

Over the next few days I will try again my AMIGA and see if the last note hangs when exiting the game, and even when AMIGA is powered on/off needing the RPi to be power cycled to stop.

**Comment by nswaldman** [comment]
Hopping on the party boat, can verify this works.

Raspi3 1.2GHz + PI-MIDI board
MacOS ScummvM, Pink Panther MID
USB UM ONE MK2

**Comment by dwhinham** [comment]
I think we can call this resolved. What a ride!

First of all, many thanks to all of you for your patience and for taking the time to test, report, and brainstorm ideas for the underlying cause.

I sincerely apologise for being suspicious of hardware when such a critical MIDI feature was missing from the software. I was so convinced of it being hardware that I couldn't see the forest for the trees. 😞 I'm still quite shocked that I'd never heard of MIDI Running Status, and a lot of sample code out there seems to lack it. But it was right there all along in the specification.

Still, it's good that popular hardware schematics have been scrutinised here, big thanks to @icb- especially for that. I will update the example schematic soon with the H11L1 alternative as it allows a simpler, safer design and avoids the possibility of underpowering the optoisolator.

You've all been given credit in the [release notes for v0.4.0](https://github.com/dwhinham/mt32-pi/releases/tag/v0.4.0), which is now available. I hope that's the end of this issue! 😄 

@Higgy69 If you still get a final hanging note in your Amiga game, please feel free to open a new issue. I think it'll be a different problem (maybe we need to implement a reset command or something), but I have Amigas over here too for testing so we can talk about it and test.

Thanks again all, and I hope you can now enjoy `mt32-pi` properly. 😄

**Comment by olliraa** [comment]
Works like a charm :D Thank you soo much @dwhinham for this lovely piece of software 🥇

**Comment by glaucon1984** [comment]
Thanks for not giving up, it's hard is to troubleshoot an issue one cannot reproduce, MT32-PI is now running perfectly for me!

**Comment by dwhinham** [comment]
My pleasure! 😄 Thanks again for your patience.

**Comment by thorr2** [comment]
I got back from my trip and just got a chance to try the new version.  I
can confirm that it mostly fixed the problem.  I don't have any further
time for testing right now, but Pink Panther sounded correct until the very
end.  At the end, the last notes hung for longer than I was expecting,
especially the cymbal drum.  There might be a problem at the end of the
song where it doesn't send the "all notes off" or something.  I only tested
the one song and didn't try it in Munt on my PC because I don't have time.

Thanks a lot for all of your help!

**Comment by thorr2** [comment]
I tried one more song and it ended properly.  I think everything is fine.
Thanks!!!

**Comment by Braincell1973** [comment]
Tested on my 2 keyboards and things much improved....

Yamaha pss680 works perfectly
Mk449 works mostly but cuts some notes short...... ( unsure if its keyboard ??? )

No change from scumm on pc , works very well.

But , As I have never owned a mt32 I cant verify this next thing , on the pc, whilst scumm seems to work properly, playing regular midi files to the mt32pi results in musical oddities , rather like its missing a track, if that makes any sense.

I can compare the output to the mt32 and Windows wave table and its like there's details missing ?? I read somewhere that the mt32 has a reserved track 0/1 ?????

Anyway , thank you for this fantastic device

**Comment by thorr2** [comment]
Braincell1973,

I haven't been following your messages, so I can't comment on the MK449 as
I am not sure how you are using it.  Regarding the last couple of
statements about playing regular MIDI files, this is normal behavior
because you are trying to play General Midi files with the MT-32, and the
MT-32 is not General MIDI compliant.  Search for Roland MT-32 MIDI files
(that have not been converted to General MIDI) and they should play
correctly.  The Pink Panther MIDI file that I linked to at the beginning of
this issue is a Roland MT-32 file.

Try these: http://ftp.monash.edu.au/pub/midi.songs/MT-32/

Hope this helps.

**Comment by dwhinham** [comment]
@thorr2 
Great to hear!

The Pink Panther MIDI has a cymbal and an open hi-hat at the end. Here are the final bars:

![image](https://user-images.githubusercontent.com/5890747/92170615-1ddd3300-ee33-11ea-9225-616ad7428689.png)

The open hi-hat sound on the MT-32 goes on for quite some time, almost 5 seconds! The cymbal has a pretty slow decay too. It sounds identical to me on both desktop Munt and `mt32-pi`, so I don't think it's anything to worry about, it's just how the MT-32's drum kit sounds.

I did some testing (both Munt and `mt32-pi`, no difference) - editing the MIDI file to add a CC 120 or 123 (All Sound Off/All Notes Off) message to the drum channel has no effect on a playing cymbal. But adding these during a long piano note cuts off the sound as expected. So these CCs, if present, do get through to the synth and work just fine, but even if they were present in the song file, it would have no effect on the drums, and this is probably just by design of the MT-32.

@Braincell1973 

Sounds good - can you check the behaviour of the Mk449 on desktop Munt when you get a chance? You can get the latest Windows version [here](https://sourceforge.net/projects/munt/files/latest/download).

Playing regular General MIDI files will rarely sound correct on an MT-32 (or an emulation of one) - the MT-32 pre-dates the General MIDI standard, so most of the instrument mappings will be wrong, and as you've noticed, channel 1 is reserved.

It's possible to get somewhat close though. On a real MT-32, holding some buttons down on startup will change the default channels from 2-9 and 10 to 1-8 and 10 (10 is always for drums). Additionally, Roland published some .MID files full of SysEx that try to remap the MT-32 instruments to resemble General MIDI.

If you grab the `MT2GM` archive [from here](http://www.houseofsynth.com/roland-mt-32-to-general-midi-conversion-utility) and play the `MTGM.MID` file to `mt32-pi`, it will remap the instruments and set the receive channels to 1-8 and 10. You won't hear anything when you play the MIDI, and it lasts around 30 seconds. After that, try playing some GM songs.

But this isn't really the intended purpose of the MT-32 - you want to find some songs composed for the MT-32 such as the ones @thorr2 has suggested. 😃

**Comment by loskaa** [comment]
@dwhinham Just wanted to let you know that after these latest updates, mt32-pi is working perfectly also with the pc-midi card (http://pcmidi.eu/). There were all sorts of strange hiccups which are completely gone with 0.5.0. Nice work!

**Comment by dwhinham** [comment]
That's what I like to hear - excellent! 🎉

Thanks for the feedback 😃
